### PR TITLE
Optimize typedesc creation for ballerina values

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/39715

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ValueUtils.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import io.ballerina.runtime.api.values.BError;
 import io.ballerina.runtime.api.values.BMap;
 import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BRefValue;
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ValueUtils.java:29]: Unnecessary import

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ValueUtils.java [hunk 2]
```diff
@@ -36,6 +37,7 @@
 import io.ballerina.runtime.internal.types.BRecordType;
 import io.ballerina.runtime.internal.values.MapValue;
 import io.ballerina.runtime.internal.values.MapValueImpl;
+import io.ballerina.runtime.internal.values.TypedescValue;
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ValueUtils.java:40]: Unused imported. There are several others too.
```
2023-02-23T14:01:11.6647396Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\ValueUtils.java:29:8: Unused import - io.ballerina.runtime.api.values.BRefValue. [UnusedImports]
2023-02-23T14:01:11.6649866Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\ValueUtils.java:40:8: Unused import - io.ballerina.runtime.internal.values.TypedescValue. [UnusedImports]
2023-02-23T14:01:11.6651645Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\values\IteratorValue.java:24:8: Unused import - io.ballerina.runtime.api.values.BTypedesc. [UnusedImports]
2023-02-23T14:01:11.6653151Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\values\RefValue.java:21:8: Unused import - io.ballerina.runtime.api.values.BTypedesc. [UnusedImports]
2023-02-23T14:01:11.6659068Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\values\StreamValue.java:32:15: Unused import - io.ballerina.runtime.internal.ValueUtils.getTypedescValue. [UnusedImports]
2023-02-23T14:01:11.6660798Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\values\TypedescValueImpl.java:20:8: Unused import - io.ballerina.runtime.api.PredefinedTypes. [UnusedImports]
2023-02-23T14:01:11.6662461Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\values\TypedescValueImpl.java:34:15: Unused import - io.ballerina.runtime.internal.ValueUtils.getTypedescValue. [UnusedImports]
2023-02-23T14:01:11.6665067Z [ant:checkstyle] [ERROR] D:\a\ballerina-lang\ballerina-lang\bvm\ballerina-runtime\src\main\java\io\ballerina\runtime\internal\values\XmlPi.java:30:15: Unused import - io.ballerina.runtime.internal.ValueUtils.getTypedescValue. [UnusedImports]
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ArrayValueImpl.java [hunk 3]
```diff
@@ -983,7 +975,6 @@ public void freezeDirect() {
                 }
             }
         }
-        this.typedesc = createSingletonTypedesc(this);
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ArrayValueImpl.java:986]: Is it okay to remove this? We can also change this to `this.typedesc = null;`

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/MapValueImpl.java [hunk 4]
```diff
@@ -530,7 +520,6 @@ public void freezeDirect() {
                 ((RefValue) val).freezeDirect();
             }
         });
-        this.typedesc = createSingletonTypedesc(this);
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/MapValueImpl.java:533]: We can set this `null` if removing this line is not okay.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java [hunk 5]
```diff
@@ -365,7 +364,6 @@ public void freezeDirect() {
         this.type = (BTableType) ReadOnlyUtils.setImmutableTypeAndGetEffectiveType(this.type);
         //we know that values are always RefValues
         this.values().forEach(val -> ((RefValue) val).freezeDirect());
-        this.typedesc = createSingletonTypedesc(this);
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java:368]: Can set this `null` if removing the line is not okay

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java [hunk 6]
```diff
@@ -594,7 +592,6 @@ public void freezeDirect() {
                 ((RefValue) value).freezeDirect();
             }
         }
-        this.typedesc = createSingletonTypedesc(this);
```
Comment (Nadeeshan96) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java:597]: Can set this `null` if removing the line is not okay. If this is not removed may need to change in subsequent places also.

**Review by hasithaa — APPROVED** [review]

**Pull request merged**